stl/inc/xstring

```diff
@@ -2527,91 +2462,83 @@ private:
 #if _HAS_CXX20
         if (_STD is_constant_evaluated()) {
             return;
         }
 #endif // _HAS_CXX20
+       // Don't annotate small strings; only annotate on the heap.
+        if (_Capacity == _BUF_SIZE - 1) {
+            return;
+        }
 
         if (!_Asan_string_should_annotate) {
             return;
         }
 
-        // We need to check whether we have a misaligned SSO buffer because of the proxy in `_Container_base` (e.g. x86)
-        if constexpr (_Memcpy_val_offset % _Asan_granularity != 0) {
-            const uintptr_t _Alignment_offset = reinterpret_cast<uintptr_t>(_Ptr) & (_Asan_granularity - 1);
-            if (_Alignment_offset != 0 && _Capacity == _BUF_SIZE - 1) {
-                return;
-            }
-        }
+        // Note that `_Capacity`, `_Old_size`, and `_New_size` do not include the null terminator
+        const void* const _End      = _First + _Capacity + 1;
+        const void* const _Old_last = _First + _Old_size + 1;
+        const void* const _New_last = _First + _New_size + 1;
 
-        // Needs to consider the null terminator
-        const char* _First    = reinterpret_cast<const char*>(_Ptr);
-        const char* _End      = reinterpret_cast<const char*>(_Ptr + _Capacity + 1);
-        const char* _Old_last = reinterpret_cast<const char*>(_Ptr + _Old_size + 1);
-        const char* _New_last = reinterpret_cast<const char*>(_Ptr + _New_size + 1);
-        if constexpr (_Has_minimum_allocation_alignment_string<basic_string>) {
-            __sanitizer_annotate_contiguous_container(_First, _End, _Old_last, _New_last);
-        } else {
-            const void* _Aligned_first = _Get_aligned_first(_First, _Capacity + 1);
-            if (!_Aligned_first) {
-                // There is no aligned address within the underlying buffer. Nothing to do
-                return;
-            }
+        constexpr bool _Large_string_always_asan_aligned =
+            (_Container_allocation_minimum_asan_alignment<basic_string>) >= _Asan_granularity;
 
-            const void* _Aligned_old_last = _Old_last < _Aligned_first ? _Aligned_first : _Old_last;
-            const void* _Aligned_new_last = _New_last < _Aligned_first ? _Aligned_first : _New_last;
-            const void* _Aligned_end      = _End < _Aligned_first ? _Aligned_first : _End;
-            __sanitizer_annotate_contiguous_container(
-                _Aligned_first, _Aligned_end, _Aligned_old_last, _Aligned_new_last);
-        }
-    }
+        // for the non-aligned buffer options, the buffer must always have size >= 9 bytes,
+        // so it will always end at least one shadow memory section.
 
-#define _ASAN_STRING_MODIFY(n)    _Modify_annotation((n))
-#define _ASAN_STRING_REMOVE(_Str) (_Str)._Remove_annotation()
-#define _ASAN_STRING_CREATE(_Str) (_Str)._Create_annotation()
+        _Asan_aligned_pointers _Aligned;
+        if constexpr (_Large_string_always_asan_aligned) {

```

Comment:
This is essentially the same as what's in vector with the SSO stuff gone, but it's written differently (clamp is avoided in vector since everything happens under the if constexpr instead of setting _Aligned). Or am I missing something?

Reply:
yeah, I could rewrite the vector stuff to more closely follow the string stuff, if you like?

Reply:
If the vector implementation is currently correct, I would prefer to see that changed in a followup PR, to avoid too much happening at once. (If the question were changing string to align more closely with vector, then that would be reasonable to do in this PR which is already changing string significantly.)